py3status/py3.py

```diff
@@ -679,3 +740,67 @@ def threshold_get_color(self, value, name=None):
         setattr(self._py3status_module, color_name, color)
 
         return color
+
+    def _get_url(self, url):
+        """
+        Get the HTTPResponse for the HTTP GET request to `url`.
+
+        Raises `Py3.URLError` on Connection Errors.
+        Raises `Py3.ClientError` on HTTP 4xx Errors.
+        Raises `Py3.ServerError` on HTTP 5xx Errors.
+        """
+        try:
+            response = HTTPResponse(urlopen(url))
+        except URLError as e:
+            self.log('Unable to access [{}]'.format(url),
+                     level=self.LOG_ERROR)
+            raise self.URLError(e)
+        status_message = '{url} returned a {http_status} status.'.format(
+            url=url, http_status=response.status)
+        if 400 <= response.status < 500:
+            self.log(status_message, level=self.LOG_ERROR)
+            raise self.ClientError(status_message)
+        elif 500 <= response.status < 600:
+            self.log(status_message, level=self.LOG_WARNING)
+            raise self.ServerError(status_message)
+        return response
+
+    def get_json_data(self, url):
+        """
+        Deserialize JSON from `url` into a Python object.
+
+        `url` can be str or a HTTPResponse object.
+
+        Returns JSON object on success, or None on error.
+        Does not raise any exceptions to caller.
+        """
+        json_data = None
+        if hasattr(url, 'getcode'):
+            response = url
+        else:
+            try:
+                response = self._get_url(url)
+            except self.HTTPError:
+                # Any Exceptions extended from `Py3.Py3Exception.HTTPError`
+                return None
+            except Exception as e:
+                self.log('An uncaught exception was raised [{}]'.format(e),
+                         level=self.LOG_ERROR)
+                return None
+        try:
+            # Using py3status.py3.HTTPResponse Object
+            json_data = response.json()
+        except:
+            # Using the built-in HTTPResponse Object
+            reader = codecs.getreader('utf-8')
+            try:
+                json_data = json.load(reader(response))
+            except (TypeError, ValueError) as e:
+                self.log('Error in json.load for [URL: {}] [{}]'.format(url, e),
+                         level=self.LOG_ERROR)
+                return None
+            except Exception as e:
+                self.log('An uncaught exception was raised [{}]'.format(e),
+                         level=self.LOG_ERROR)
+                return None
```

Comment:
Generally, module authors should use `get_json_data()`, which will mask all errors with `None`.  If someone is working with the `HTTPResponse` object directly, I imagine that any exceptions should not be hidden.

I had the logic of trying the `.json()` method, then if it failed for any reason, fall back to the original method.  This would catch someone passing in the output of `urlopen()` and also dealing with invalid JSON.  With removing the pass through API, this would not work with invalid JSON.

I'll clean this up a bit more.